File: backend/src/main/java/com/autocare/serviceorder/service/ServiceOrderService.java
package com.autocare.serviceorder.service;

import com.autocare.auth.entity.Role;
import com.autocare.auth.entity.User;
import com.autocare.client.entity.Client;
import com.autocare.client.service.ClientService;
import com.autocare.mechanic.entity.Mechanic;
import com.autocare.mechanic.repository.MechanicRepository;
import com.autocare.mechanic.service.MechanicService;
import com.autocare.serviceorder.dto.ServiceOrderRequestDTO;
import com.autocare.serviceorder.dto.ServiceOrderResponseDTO;
import com.autocare.serviceorder.dto.StatusUpdateRequestDTO;
import com.autocare.serviceorder.entity.ServiceOrder;
import com.autocare.serviceorder.entity.ServiceOrderStatus;
import com.autocare.serviceorder.repository.ServiceOrderRepository;
import com.autocare.shared.exception.BusinessException;
import com.autocare.shared.exception.ResourceNotFoundException;
import com.autocare.shared.security.SecurityUtils;
import com.autocare.vehicle.entity.Vehicle;
import com.autocare.vehicle.service.VehicleService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ServiceOrderService {

    private final ServiceOrderRepository serviceOrderRepository;
    private final ClientService clientService;
    private final VehicleService vehicleService;
    private final MechanicService mechanicService;
    private final MechanicRepository mechanicRepository;

    @Transactional
    public ServiceOrderResponseDTO create(ServiceOrderRequestDTO request) {
        Client client = clientService.findById(request.getClientId());
        Vehicle vehicle = vehicleService.findById(request.getVehicleId());
        Mechanic mechanic = request.getMechanicId() != null ?
                mechanicService.findById(request.getMechanicId()) : null;

        String orderNumber = generateOrderNumber();

        ServiceOrder serviceOrder = ServiceOrder.builder()
                .orderNumber(orderNumber)
                .client(client)
                .vehicle(vehicle)
                .mechanic(mechanic)
                .reportedProblem(request.getReportedProblem())
                .status(ServiceOrderStatus.CRIADA)
                .totalAmount(java.math.BigDecimal.ZERO)
                .build();

        serviceOrder = serviceOrderRepository.save(serviceOrder);
        return toResponseDTO(serviceOrder);
    }

    @Transactional
    public ServiceOrderResponseDTO updateStatus(String id, StatusUpdateRequestDTO request) {
        ServiceOrder serviceOrder = findById(id);
        assertCanAccessOrder(serviceOrder);

        validateStatusTransition(serviceOrder.getStatus(), request.getStatus());

        if (request.getDiagnosis() != null) {
            serviceOrder.setDiagnosis(request.getDiagnosis());
        }

        serviceOrder.setStatus(request.getStatus());

        if (request.getStatus() == ServiceOrderStatus.EM_EXECUCAO) {
            serviceOrder.setStartedAt(LocalDateTime.now());
        }

        if (request.getStatus() == ServiceOrderStatus.FINALIZADA) {
            serviceOrder.setCompletedAt(LocalDateTime.now());
        }

        serviceOrder = serviceOrderRepository.save(serviceOrder);
        return toResponseDTO(serviceOrder);
    }

    @Transactional
    public ServiceOrderResponseDTO assignMechanic(String id, String mechanicId) {
        ServiceOrder serviceOrder = findById(id);
        Mechanic mechanic = mechanicService.findById(mechanicId);

        serviceOrder.setMechanic(mechanic);
        serviceOrder = serviceOrderRepository.save(serviceOrder);
        return toResponseDTO(serviceOrder);
    }

    public ServiceOrderResponseDTO findByIdResponse(String id) {
        ServiceOrder serviceOrder = findById(id);
        assertCanAccessOrder(serviceOrder);
        return toResponseDTO(serviceOrder);
    }

    public List<ServiceOrderResponseDTO> findAll() {
        return serviceOrderRepository.findAll()
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<ServiceOrderResponseDTO> findByClient(String clientId) {
        return serviceOrderRepository.findByClientId(clientId)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<ServiceOrderResponseDTO> findByVehicle(String vehicleId) {
        return serviceOrderRepository.findByVehicleId(vehicleId)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<ServiceOrderResponseDTO> findByStatus(ServiceOrderStatus status) {
        return serviceOrderRepository.findByStatus(status)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public ServiceOrder findById(String id) {
        return serviceOrderRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Ordem de serviço não encontrada com ID: " + id));
    }

    /**
     * Restringe o acesso de um usuário MECHANIC apenas às ordens de serviço que lhe
     * foram atribuídas — o controller libera a rota para o papel MECHANIC (ele precisa
     * ver e diagnosticar suas OS), mas sem esta checagem qualquer mecânico logado
     * conseguiria ler ou alterar a OS de qualquer colega. Outros papéis (ADMIN,
     * MANAGER, RECEPTIONIST) não passam por essa restrição adicional.
     */
    private void assertCanAccessOrder(ServiceOrder serviceOrder) {
        User currentUser = SecurityUtils.getCurrentUser();
        if (currentUser.getRole() != Role.MECHANIC) {
            return;
        }

        Mechanic mechanic = mechanicRepository.findByUserId(currentUser.getId())
                .orElseThrow(() -> new AccessDeniedException("Usuário não possui um perfil de mecânico vinculado"));

        boolean isAssigned = serviceOrder.getMechanic() != null
                && serviceOrder.getMechanic().getId().equals(mechanic.getId());

        if (!isAssigned) {
            throw new AccessDeniedException("Esta ordem de serviço não está atribuída a você");
        }
    }

    private void validateStatusTransition(ServiceOrderStatus current, ServiceOrderStatus newStatus) {
        if (current == newStatus) {
            return;
        }

        boolean isValid = switch (current) {
            case CRIADA -> newStatus == ServiceOrderStatus.EM_DIAGNOSTICO || newStatus == ServiceOrderStatus.CANCELADA;
            case EM_DIAGNOSTICO -> newStatus == ServiceOrderStatus.AGUARDANDO_APROVACAO || newStatus == ServiceOrderStatus.CANCELADA;
            case AGUARDANDO_APROVACAO -> newStatus == ServiceOrderStatus.APROVADA || newStatus == ServiceOrderStatus.CANCELADA;
            case APROVADA -> newStatus == ServiceOrderStatus.EM_EXECUCAO || newStatus == ServiceOrderStatus.CANCELADA;
            case EM_EXECUCAO -> newStatus == ServiceOrderStatus.FINALIZADA || newStatus == ServiceOrderStatus.CANCELADA;
            case FINALIZADA, CANCELADA -> false;
        };

        if (!isValid) {
            throw new BusinessException(
                    String.format("Transição de status inválida de %s para %s", current, newStatus)
            );
        }
    }

    private String generateOrderNumber() {
        String prefix = "OS";
        String year = String.valueOf(LocalDateTime.now().getYear());
        // nextval() é atômico no Postgres: cada chamada concorrente recebe um valor
        // diferente e garantido único, ao contrário de "count() + 1" (ver repository).
        String sequence = String.format("%06d", serviceOrderRepository.nextOrderSequenceValue());
        return prefix + year + sequence;
    }

    private ServiceOrderResponseDTO toResponseDTO(ServiceOrder serviceOrder) {
        return ServiceOrderResponseDTO.builder()
                .id(serviceOrder.getId())
                .orderNumber(serviceOrder.getOrderNumber())
                .clientId(serviceOrder.getClient().getId())
                .clientName(serviceOrder.getClient().getName())
                .vehicleId(serviceOrder.getVehicle().getId())
                .vehicleInfo(serviceOrder.getVehicle().getBrand() + " " +
                        serviceOrder.getVehicle().getModel() + " (" +
                        serviceOrder.getVehicle().getPlate() + ")")
                .mechanicId(serviceOrder.getMechanic() != null ? serviceOrder.getMechanic().getId() : null)
                .mechanicName(serviceOrder.getMechanic() != null ? serviceOrder.getMechanic().getName() : null)
                .reportedProblem(serviceOrder.getReportedProblem())
                .diagnosis(serviceOrder.getDiagnosis())
                .status(serviceOrder.getStatus())
                .totalAmount(serviceOrder.getTotalAmount())
                .startedAt(serviceOrder.getStartedAt())
                .completedAt(serviceOrder.getCompletedAt())
                .createdAt(serviceOrder.getCreatedAt())
                .build();
    }
}

File: backend/src/service-order/service-order.service.ts
import { ForbiddenException, Injectable } from "@nestjs/common";
import { Role, ServiceOrderStatus } from "@prisma/client";
import { PrismaService } from "../prisma/prisma.service";
import { BusinessException, ResourceNotFoundException } from "../common/exceptions";
import { ClientService } from "../client/client.service";
import { VehicleService } from "../vehicle/vehicle.service";
import { MechanicService } from "../mechanic/mechanic.service";
import { ServiceOrderRequestDto, StatusUpdateRequestDto } from "./dto/service-order.dto";
import { AuthUser } from "../auth/jwt.strategy";

// Espelha ServiceOrderService.validateStatusTransition do backend original.
const VALID_TRANSITIONS: Record<ServiceOrderStatus, ServiceOrderStatus[]> = {
  CRIADA: [ServiceOrderStatus.EM_DIAGNOSTICO, ServiceOrderStatus.CANCELADA],
  EM_DIAGNOSTICO: [ServiceOrderStatus.AGUARDANDO_APROVACAO, ServiceOrderStatus.CANCELADA],
  AGUARDANDO_APROVACAO: [ServiceOrderStatus.APROVADA, ServiceOrderStatus.CANCELADA],
  APROVADA: [ServiceOrderStatus.EM_EXECUCAO, ServiceOrderStatus.CANCELADA],
  EM_EXECUCAO: [ServiceOrderStatus.FINALIZADA, ServiceOrderStatus.CANCELADA],
  FINALIZADA: [],
  CANCELADA: [],
};

@Injectable()
export class ServiceOrderService {
  constructor(
    private prisma: PrismaService,
    private clientService: ClientService,
    private vehicleService: VehicleService,
    private mechanicService: MechanicService,
  ) {}

  async create(dto: ServiceOrderRequestDto) {
    const client = await this.clientService.findById(dto.clientId);
    const vehicle = await this.vehicleService.findById(dto.vehicleId);
    if (dto.mechanicId) {
      await this.mechanicService.findById(dto.mechanicId);
    }

    const orderNumber = await this.generateOrderNumber();

    const order = await this.prisma.serviceOrder.create({
      data: {
        orderNumber,
        clientId: client.id,
        vehicleId: vehicle.id,
        mechanicId: dto.mechanicId,
        reportedProblem: dto.reportedProblem,
        status: ServiceOrderStatus.CRIADA,
        totalAmount: 0,
      },
      include: { client: true, vehicle: true, mechanic: true },
    });
    return this.toResponse(order);
  }

  async updateStatus(id: string, dto: StatusUpdateRequestDto, currentUser: AuthUser) {
    const order = await this.findByIdWithRelations(id);
    await this.assertCanAccessOrder(order, currentUser);

    this.validateStatusTransition(order.status, dto.status);

    const data: any = { status: dto.status };
    if (dto.diagnosis != null) {
      data.diagnosis = dto.diagnosis;
    }
    if (dto.status === ServiceOrderStatus.EM_EXECUCAO) {
      data.startedAt = new Date();
    }
    if (dto.status === ServiceOrderStatus.FINALIZADA) {
      data.completedAt = new Date();
    }

    const updated = await this.prisma.serviceOrder.update({
      where: { id },
      data,
      include: { client: true, vehicle: true, mechanic: true },
    });
    return this.toResponse(updated);
  }

  async assignMechanic(id: string, mechanicId: string) {
    await this.findById(id);
    await this.mechanicService.findById(mechanicId);

    const updated = await this.prisma.serviceOrder.update({
      where: { id },
      data: { mechanicId },
      include: { client: true, vehicle: true, mechanic: true },
    });
    return this.toResponse(updated);
  }

  async findByIdResponse(id: string, currentUser: AuthUser) {
    const order = await this.findByIdWithRelations(id);
    await this.assertCanAccessOrder(order, currentUser);
    return this.toResponse(order);
  }

  async findAll() {
    const orders = await this.prisma.serviceOrder.findMany({
      include: { client: true, vehicle: true, mechanic: true },
      orderBy: { createdAt: "desc" },
    });
    return orders.map((o) => this.toResponse(o));
  }

  async findByClient(clientId: string) {
    const orders = await this.prisma.serviceOrder.findMany({
      where: { clientId },
      include: { client: true, vehicle: true, mechanic: true },
    });
    return orders.map((o) => this.toResponse(o));
  }

  async findByVehicle(vehicleId: string) {
    const orders = await this.prisma.serviceOrder.findMany({
      where: { vehicleId },
      include: { client: true, vehicle: true, mechanic: true },
    });
    return orders.map((o) => this.toResponse(o));
  }

  async findByStatus(status: ServiceOrderStatus) {
    const orders = await this.prisma.serviceOrder.findMany({
      where: { status },
      include: { client: true, vehicle: true, mechanic: true },
    });
    return orders.map((o) => this.toResponse(o));
  }

  // Retorna a entidade crua — usado internamente por BudgetService.
  async findById(id: string) {
    const order = await this.prisma.serviceOrder.findUnique({ where: { id } });
    if (!order) {
      throw new ResourceNotFoundException(`Ordem de serviço não encontrada com ID: ${id}`);
    }
    return order;
  }

  private async findByIdWithRelations(id: string) {
    const order = await this.prisma.serviceOrder.findUnique({
      where: { id },
      include: { client: true, vehicle: true, mechanic: true },
    });
    if (!order) {
      throw new ResourceNotFoundException(`Ordem de serviço não encontrada com ID: ${id}`);
    }
    return order;
  }

  // Restringe o acesso de um usuário MECHANIC apenas às ordens de serviço que lhe
  // foram atribuídas — o controller libera a rota para o papel MECHANIC (ele precisa
  // ver e diagnosticar suas OS), mas sem esta checagem qualquer mecânico logado
  // conseguiria ler ou alterar a OS de qualquer colega. Outros papéis não passam por
  // essa restrição adicional.
  private async assertCanAccessOrder(order: { mechanicId: string | null }, currentUser: AuthUser) {
    if (currentUser.role !== Role.MECHANIC) {
      return;
    }

    const mechanic = await this.mechanicService.findByUserId(currentUser.id);
    if (!mechanic) {
      throw new ForbiddenException("Usuário não possui um perfil de mecânico vinculado");
    }

    const isAssigned = order.mechanicId != null && order.mechanicId === mechanic.id;
    if (!isAssigned) {
      throw new ForbiddenException("Esta ordem de serviço não está atribuída a você");
    }
  }

  private validateStatusTransition(current: ServiceOrderStatus, next: ServiceOrderStatus) {
    if (current === next) {
      return;
    }
    const allowed = VALID_TRANSITIONS[current] ?? [];
    if (!allowed.includes(next)) {
      throw new BusinessException(`Transição de status inválida de ${current} para ${next}`);
    }
  }

  // Usa um contador atômico (tabela Counter, incrementado dentro de uma transação)
  // em vez de "count() + 1": count()+1 não é atômico, então duas OS criadas ao mesmo
  // tempo podiam calcular o mesmo próximo número. O UPDATE...increment é garantido
  // atômico pelo Postgres mesmo sob concorrência — mesma garantia que o nextval()
  // da sequence usada no backend original.
  private async generateOrderNumber(): Promise<string> {
    const counter = await this.prisma.counter.upsert({
      where: { name: "service_order" },
      update: { value: { increment: 1 } },
      create: { name: "service_order", value: 1 },
    });
    const year = new Date().getFullYear();
    const sequence = String(counter.value).padStart(6, "0");
    return `OS${year}${sequence}`;
  }

  private toResponse(order: any) {
    return {
      id: order.id,
      orderNumber: order.orderNumber,
      clientId: order.clientId,
      clientName: order.client?.name,
      vehicleId: order.vehicleId,
      vehicleInfo: order.vehicle ? `${order.vehicle.brand} ${order.vehicle.model} (${order.vehicle.plate})` : undefined,
      mechanicId: order.mechanicId,
      mechanicName: order.mechanic?.name,
      reportedProblem: order.reportedProblem,
      diagnosis: order.diagnosis,
      status: order.status,
      totalAmount: order.totalAmount,
      startedAt: order.startedAt,
      completedAt: order.completedAt,
      createdAt: order.createdAt,
    };
  }
}
